### PR TITLE
ADNI MRI brain image super-resolution network using sub-pixel CNN

### PR DESCRIPTION
**Pull request by DHyunC**
# ADNI MRI brain image super-resolution network using sub-pixel CNN
## Purpose
This pull request presents a super-resolution network which aims to increase the resolution of the input 2d brain MRI image by a factor of 4.

## General Overview
Super-resolution convolutional neural networks (SRCNN) are deep learning models which aims to increase the quality and resolution of images by learning the mapping between low quality and the high quality images. They are often used in MRI, CT scans and even in satellite imagery. 

## Core Functionality
The core component of this network is the sub-pixel layer which utilizes the ``nn.PixelShuffle`` function from pytorch to compute efficient sub-pixel upscaling by a given factor. 

## Adjustments and optimizations
When the files are ran without adjusting the epochs or the hyper-parameters, the results should reproduce the following:
Loss: 0.00173 (may fluctuate)
It should be noted that even with much lower epochs (default is 100) such as 5 epochs, similar results could be achieved. 5 epochs should lead to a loss of about 0.00203. Also, having low epochs results in a drastically faster computation time as 5 epochs only took 3.8 minutes on a 30 series Nvidia GPU. The computation time can be further decreased by increasing batch size from the default of 30.

## Personal takeaway
Overall, this project has opened my eye as to how real life applications of neural networks can positively affect others and I found the implementation of the network to be enjoyable.

Daniel Choi

### CODE REVIEW

**Comment by uqmarlonbran** [comment]
**This is an initial inspection - no action is required yet**

**Difficulty:** Normal

**General Comments:** Seems to provide a nice and working implementation of PixelCNN and its associated training procedure. Utils.py is used to change the data directories, as well as where the model is loaded from after training. Does not implement a validation set and does not provide a mechanism to check performance on the testing dataset. Does provide some ablation studies regarding filter size and choice of activation function which is nice to see.

**[Recognition Problem]**
"modules.py":

- Correct implementation of PixelCNN, no changes architecturally but allows for variable upscaling factor and number of filter channels. Though, number of channels is adjusted from the utils file rather than the init of the model class itself.

"dataset.py":

- Nice implementation using ImageFolder and DataLoader to collect images from AD and NC folders.

"train.py":

- Nice training loop, however, no validation set split.
- Accumulates the training loss at the end of each epoch and reports it correctly in the provided figure.
- Simply saves the latest version of the model (rather than checking for lowest validation loss).

"predict.py":

- Correctly loads model from saved weights and tests of sample images.
- Simply iterates through each image in the test and displays the reconstruction (no metrics recorded).

"README.MD":

- README provides an overview of the network and is well laid out. Also provides some example inputs and outputs. Some contrast seems to have changed. Lists dependencies and provides a training plot (though no validation). Does not provide performance on test dataset.

**[Commit Log]**
Seems to have a thorough commit log.

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks missing -1
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Normal -5
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe good
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
